analyzer/src/main/kotlin/AnalyzerResultBuilder.kt

```diff
@@ -46,9 +46,18 @@ class AnalyzerResultBuilder {
 
     fun build(excludes: Excludes = Excludes.EMPTY): AnalyzerResult {
         val duplicates = (projects.map { it.toPackage() } + packages).getDuplicates { it.id }
-        require(duplicates.isEmpty()) {
+
+        // Some source code repositories contain projects that are used as packages by other contained projects. So the
+        // same code is once seen as a project, and once as a package by ORT. Allow such "duplicates" as the ids
+        // actually refer to the same thing by requiring duplicates to refer to different VCS locations.
+        val realDuplicates = duplicates.filter { (_, duplicates) ->
```

Comment:
@oss-review-toolkit/core-devs, what do you thing about this? Should we rather only keep one of these (either the project or the package) instead of keeping both of them?

Reply:
I'm specifically asking as I currently don't oversee all side-effects that might occur if we keep a project and a package with the same id in the result.

Reply:
I think it's likely that assumption that Id's are unique are spread all over the place.  To list a few:

- https://github.com/oss-review-toolkit/ort/blob/19c89ff9a0a7aa2a52d85c82fd477531da1ecf3d/model/src/main/kotlin/OrtResult.kt#L451
- https://github.com/oss-review-toolkit/ort/blob/19c89ff9a0a7aa2a52d85c82fd477531da1ecf3d/model/src/main/kotlin/OrtResult.kt#L532
- https://github.com/oss-review-toolkit/ort/blob/19c89ff9a0a7aa2a52d85c82fd477531da1ecf3d/helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt#L66
- https://github.com/oss-review-toolkit/ort/blob/main/model/src/main/kotlin/OrtResult.kt#L132

Reply:
> Right. But some of these only assume ids to be unique within projects _or_ packages, but not within the union of both.

Yeah, that's for sure.

Reply:
Have you considered as alternative solution: Making sure that the `Identifier.type` for projects differs from `Identifier.type` for packages?

Reply:
ISTR that there was also a problem with the WebApp report when duplicate IDs were present: The JavaScript part uses an index-based mapping to reference objects, and if there were projects and packages with the same ID, this mapping got broken.

Reply:
> Have you considered as alternative solution: Making sure that the `Identifier.type` for projects differs from `Identifier.type` for packages?

Actually, I was playing around with similar ideas 3 years (!) ago:
- In the [identifier-types](https://github.com/oss-review-toolkit/ort/compare/main...identifier-types) branch I tried to turn an `Identifier`'s `type` into a `sealed class` with `ProjectIdentifier` and `PackageIdentifier` sub-classes.
- In the [multiple-identifier-classes](https://github.com/oss-review-toolkit/ort/compare/main...multiple-identifier-classes) branch the `Identifier` class itself is tried to be turned in a `sealed class`.

Both approaches had the goal to distinguish project and package ids, and to iterate over all known id types (e.g. for `when` statements).

Reply:
I agree with @fviernau that there are probably lots of places in the code which assume that identifiers are unique and it could be very hard to find all of them, and with @oheger-bosch that likely the WebApp report and probably other report formats would break.
There are also situations where the same package could appear twice with different metadata (e.g. if different repositories are used), I'm not sure if we currently fail in this case or just ignore it.
I think properly addressing this issue would be a big effort, another approach than the sealed class you suggested could also be to extend the identifier with properties like `isProject` and provider details.

Reply:
> I think properly addressing this issue would be a big effort

Indeed. That's why I never completed my branches mentioned above...